contracts/Gauge.sol

```diff
@@ -383,162 +435,246 @@ contract Gauge is BaseGauge, IGauge {
+        //take away from sender
+        asset.safeTransferFrom(msg.sender, address(this), _assets);
 
-        _totalSupply += _amount;
-        uint256 newBalance = balance.realBalance + _amount;
-        balance.realBalance = newBalance;
-        uint256 boostedBalance = _boostedBalanceOf(_for, newBalance);
-        balance.boostedBalance = boostedBalance;
-        emit BoostedBalanceUpdated(_for, boostedBalance);
+        // mint shares
+        _mint(_receiver, _assets);
 
-        //take away from sender
-        stakingToken.safeTransferFrom(msg.sender, address(this), _amount);
-        emit Staked(_for, _amount);
+        uint256 boostedBalance = _boostedBalanceOf(_receiver);
+        _boostedBalances[_receiver] = boostedBalance;
+
+        emit BoostedBalanceUpdated(_receiver, boostedBalance);
+        emit Deposit(msg.sender, _receiver, _assets, _assets);
+    }
+
+    /**
+      Maximum amount of the underlying asset that can be withdrawn from the owner balance in the Vault, through a withdraw call.
+    */
+    function maxWithdraw(address _owner) external view returns (uint256) {
+        return balanceOf(_owner);
+    }
+
+    function previewWithdraw(uint256 _assets) external view returns (uint256) {
+        return _assets;
     }
 
     /** @notice allow an address to deposit on your behalf
      *  @param _addr address to change approval for
-     *  @param _canDeposit can deposit
      *  @param _canClaim can deposit
      *  @return true
      */
     function setApprovals(
         address _addr,
-        bool _canDeposit,
         bool _canClaim,
         bool _canLock
     ) external returns (bool) {
-        approvedTo[_addr][msg.sender].deposit = _canDeposit;
         approvedTo[_addr][msg.sender].claim = _canClaim;
         approvedTo[_addr][msg.sender].lock = _canLock;
 
         return true;
     }
 
-    /** @notice withdraw vault token from the gauge
+    /** @notice Burns shares from owner and sends exactly assets of underlying tokens to receiver.
      *  @dev This call updates claimable rewards
-     *  @param _amount amount to withdraw
-     *  @param _claim claim veYFI and additional reward
+     *  @param _assets amount to withdraw
+     *  @param _receiver account that will recieve the shares
+     *  @param _owner shares will be taken from account
      *  @param _lock should the claimed rewards be locked in veYFI for the user
-     *  @return true
+     *  @param _claim claim veYFI and additional reward
+     *  @return amount of shares withdrawn
      */
     function withdraw(
-        uint256 _amount,
+        uint256 _assets,
+        address _receiver,
+        address _owner,
         bool _claim,
         bool _lock
-    ) public updateReward(msg.sender) returns (bool) {
-        require(_amount != 0, "RewardPool : Cannot withdraw 0");
-        Balance storage balance = _balances[msg.sender];
-        require(
-            balance.lastDeposit < block.number,
```

Comment:
Cant we add this balance.lastDeposit check state on transfers too? that way we don't need to remove it here since its a valid guard, as pointed out from yacademy

Reply:
Not sure the kick function is a complete mitigation to the flashloan boost problem, its similar to the harvest sandwiches in vault and before we had lockedProfitDegradation it was a complete whack a mole game. 

Is the plan to mitigate when we send rewards to kick in the same trx? i can see us getting sandwiched when we drop rewards or after kicking someone, they can flashloan again and boost after in different account after transfering.

Basically im not sure it can be solve with lastDeposit check, may need to have degradation on deposits in the same way harvest profit also has locking and degradation.

Reply:
It will be an issue for contracts that are depositing and transferring funds to an EOA (zap)

Reply:
> It will be an issue for contracts that are depositing and transferring funds to an EOA (zap)

you mean to track lastDeposit, yeah i don't see it as a fully complete solution because of the DDOS problem, if we don't allow users to withdraw based on lastDeposit someone can flood you with dust your account to front run lastDeposit and keep you from withdrawing, still not sure about another alternative.